boofuzz/socket_connection.py

```diff
@@ -159,12 +159,28 @@ def open(self):
 
         if self.server:
             self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            self._sock.bind((self.host, self.port))
+            try:
+                self._sock.bind((self.host, self.port))
+            except socket.error as e:
+                # Handle when socket is in TIME_WAIT state.
+                if e.errno == errno.EADDRINUSE:
+                    raise exception.BoofuzzOutOfAvailableSockets()
+                else:
+                    raise
 
             if self.proto == "tcp" or self.proto == "ssl":
                 self._serverSock = self._sock
-                self._serverSock.listen(1)
-                self._sock, addr = self._serverSock.accept()
+                try:
+                    self._serverSock.listen(1)
+                    self._sock, addr = self._serverSock.accept()
+                except socket.error as e:
+                    # When connection timeout expires, tear down the server socket so we can re-open it again after
+                    # restarting the target. Can't just call self.close() because self._sock isn't as expected for it.
+                    self._serverSock.close()
+                    if e.errno in [errno.EAGAIN]:
```

Comment:
When the process cannot allocate a new socket (for example because of `TIME_WAIT` problem), the error number in the socket is `errno.EADDRINUSE`. Or is this behaviour different for `bind` method? If not, I would suggest handling `errno.EADDRINUSE` too.

Like:
```python
if e.errno in [errno.EAGAIN, errno.EADDRINUSE]:
    raise exception.BoofuzzTargetConnectionFailedError(str(e))
```

Or maybe I just missing something. Feel free to correct me, I am quite a beginner in this field. Thanks.